src/org/mozilla/javascript/Kit.java

```diff
@@ -144,6 +197,52 @@ public static int xDigitToInt(int c, int accumulator)
         return (accumulator << 4) | c;
     }
 
+    public static void reinitNonClassCache() {
+        notClassMap = Collections.synchronizedMap(new WeakHashMap<ClassLoader, LRUCache>());
+    }
+
+    private static synchronized LRUCache getOrCreateCache(ClassLoader loader) {
+        LRUCache lruCache = notClassMap.get(loader);
+        if (lruCache == null) {
+            lruCache = new LRUCache(getDefaultNonClassCacheSize());
+            notClassMap.put(loader, lruCache);
+        }
+        return lruCache;
+    }
+
+    private static int getDefaultNonClassCacheSize() {
+        try {
+            String propValue = System.getProperty(CACHE_USE_PROPERTY_NAME);
+            if (propValue != null && propValue.length() != 0)
+                return Integer.valueOf(propValue);
+            else return MAX_CAPACITY_DEFAULT;
+        } catch (Exception e) {
+            e.printStackTrace(System.err);
+            return MAX_CAPACITY_DEFAULT;
+        }
+
+    }
+
+    public static int getNonClassCacheSize() {
+        int result = 0;
+
+        for (Map.Entry<ClassLoader, LRUCache> entry : notClassMap.entrySet()) {
+            if (entry instanceof WeakReference) {
+                WeakReference<ClassLoader> weak = (WeakReference<ClassLoader>)entry;
+                if (!weak.isEnqueued())
+                    result += entry.getValue().size();
+            } else {
+                System.err.println("strange entry type");
```

Comment:
I think that if this is really something that should never happen then you should throw an AssertionError here and be done with it.

OTOH if it's something that we need to recover from then you should return a reasonable result.

Otherwise, having Rhino print to stderr doesn't really help anyone and will just cause us problems later.

Also this isn't thread-safe ;-)